### PR TITLE
[ML Agent] Add daemon process code for ML Service

### PR DESCRIPTION
**Pull request by again4you**
This patch newly adds the daemon process for ML Service. It provides a 
basic template code so developers can add DBus interface and its service code.

Signed-off-by: Sangjung Woo <sangjung.woo@samsung.com>

### Component Diagram
uml_01

#### V3
* Add systemd service file without target link.
* Fix the style issue such as unnecessary new line, typoes, wrong indentation.
* Use `g_clear_object()` instead of `g_object_unref()` and seting NULL.

#### V2
* Make 5 components to improve maintainability
* Developers can easily add `DBus interface` by implementing its skeleton code.

#### V1
* Make a Draft version for testing on Tizen Device and sharing current status

### CODE REVIEW

File: c/meson.build [hunk 1]
```diff
@@ -213,3 +213,5 @@ configure_file(input: 'capi-ml-common.pc.in', output: 'capi-ml-common.pc',
 install_headers(nns_capi_headers,
   subdir: 'nnstreamer'
 )
+
+
```
Comment (myungjoo) [inline c/meson.build:217]: useless lines

File: daemon/main.c [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+#include <glib.h>
+#include <gio/gio.h>
+
+#include "common.h"
+#include "modules.h"
+#include "gdbus-util.h"
+#include "log.h"
+#include "dbus-interface.h"
+
+GMainLoop *g_mainloop;
```
Comment (myungjoo) [inline daemon/main.c:24]: static?

**Review by myungjoo — APPROVED** [review]
LGTM other than minor style issues.

File: daemon/gdbus-util.c [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+
+/**
+ * @brief Connect to the DBus message bus, which type is SYSTEM.
+ */
+int
+gdbus_get_system_connection (void)
+{
+  GError *error = NULL;
+
+  g_dbus_sys_conn = g_bus_get_sync (G_BUS_TYPE_SYSTEM, NULL, &error);
```
Comment (jihochu) [inline daemon/gdbus-util.c:153]: It's well, but how about using async api `g_buf_get`.
Most of gdbus methods are called by asynchrously, thus you can handle proper dbus init timing with g_main_loop_run.

File: daemon/gdbus-util.c [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+}
+
+/**
+ * @brief Disconnect the DBus message bus.
+ */
+void
+gdbus_put_system_connection (void)
+{
+  if (g_dbus_sys_conn) {
+    g_object_unref (g_dbus_sys_conn);
```
Comment (jihochu) [inline daemon/gdbus-util.c:169]: This work well, but how about using `g_clear_object`  for convenience ?
https://docs.gtk.org/gobject/func.clear_object.html

**Comment by jihochu** [comment]
Compnent Diagram really good for understadning :+1:

File: daemon/gdbus-util.c [hunk 5]
```diff
@@ -0,0 +1,172 @@
+/* SPDX-License-Identifier: Apache-2.0 */
+/**
+ * Copyright (c) 2022 Samsung Electronics Co., Ltd. All Rights Reserved.
+ *
+ * @file gdbus-util.c
+ * @date 25 June 2022
+ * @brief Internal GDbus utility warpper of Machine Learning agent daemon
```
Comment (wooksong) [inline daemon/gdbus-util.c:7]:
```suggestion
 * @brief Internal GDbus utility wrapper of Machine Learning agent daemon
```

File: daemon/meson.build [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+if get_option('enable-machine-learning-agent')
+  nns_ml_agent_srcs = []
+  nns_ml_agent_incs = []
+  nns_ml_agent_incs += include_directories('includes')
+
+  # Generate GDbus header and code
+  gdbus_prog = find_program('gdbus-codegen')
+  if not gdbus_prog.found()
+    error('gdbus_codegen is not found!')
+  endif
```
Comment (wooksong) [inline daemon/meson.build:11]:
```suggestion
  gdbus_prog = find_program('gdbus-codegen', required : true)
```

File: daemon/meson.build [hunk 7]
```diff
@@ -0,0 +1,59 @@
+# Machine Learing Agent
+if get_option('enable-machine-learning-agent')
+  nns_ml_agent_srcs = []
+  nns_ml_agent_incs = []
+  nns_ml_agent_incs += include_directories('includes')
```
Comment (wooksong) [inline daemon/meson.build:5]:
```suggestion
  nns_ml_agent_incs = include_directories('includes')
```

File: meson.build [hunk 8]
```diff
@@ -153,6 +157,10 @@ api_conf.set('INCLUDE_INSTALL_DIR', api_install_includedir)
 # Build C-API
 subdir('c')
 
+if get_option('enable-machine-learning-agent')
+subdir('daemon')
```
Comment (wooksong) [inline meson.build:161]:
```suggestion
  subdir('daemon')
```

File: packaging/machine-learning-api.spec [hunk 9]
```diff
@@ -287,6 +307,7 @@ HTML pages of lcov results of ML API generated during rpm build
 %prep
 %setup -q
 cp %{SOURCE1001} .
+cp %{SOURCE1002} .
```
Comment (wooksong) [inline packaging/machine-learning-api.spec:310]: Is ```%if 0%{?enable_machine_learning_agent}``` unnecessary here?

File: c/meson.build (hunk 1 above)
Comment (again4you) [reply c/meson.build:217]: Fixed!

File: daemon/main.c (hunk 2 above)
Comment (again4you) [reply daemon/main.c:24]: Right. updated!

File: daemon/gdbus-util.c (hunk 5 above)
Comment (again4you) [reply daemon/gdbus-util.c:7]: Applied! Thanks.

File: daemon/meson.build (hunk 6 above)
Comment (again4you) [reply daemon/meson.build:11]: Simple one line is easy to understand. Applied!

File: daemon/meson.build (hunk 7 above)
Comment (again4you) [reply daemon/meson.build:5]: Applied!

File: meson.build (hunk 8 above)
Comment (again4you) [reply meson.build:161]: Applied!

File: packaging/machine-learning-api.spec (hunk 9 above)
Comment (again4you) [reply packaging/machine-learning-api.spec:310]: You're right. My fault!.

File: meson.build (hunk 8 above)
Comment (again4you) [reply meson.build:161]: Applied!

File: daemon/meson.build (hunk 7 above)
Comment (again4you) [reply daemon/meson.build:5]: Applied!

File: daemon/gdbus-util.c [hunk 10]
```diff
@@ -0,0 +1,172 @@
...
+/**
+ * @brief Disconnect the DBus message bus.
+ */
+void
+gdbus_put_system_connection (void)
+{
+  if (g_dbus_sys_conn) {
+    g_object_unref (g_dbus_sys_conn);
+    g_dbus_sys_conn = NULL;
+  }
```
Comment (again4you) [inline daemon/gdbus-util.c:171]: The function you recommended is much clear than the previous logic. I will fix it as below.

```suggestion
  g_clear_object (&g_dbus_sys_conn);
```

File: daemon/gdbus-util.c (hunk 3 above)
Comment (again4you) [reply daemon/gdbus-util.c:153]: This is the first draft for ML agent daemon and the asynchrous method is good enough for our requirements. Next time I will consider the async API to improve the performance.

**Comment by again4you** [comment]
#### V3
* Add systemd service file without target link.
* Fix the style issue such as unnecessary new line, typoes, wrong indentation.
* Use `g_clear_object()` instead of `g_object_unref()` and seting NULL.

**Review by wooksong — APPROVED** [review]
LGTM

**Pull request merged**